enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -1934,15 +1884,106 @@ func (c *FirecrackerContainer) Pause(ctx context.Context) error {
 	defer func() {
 		log.CtxDebugf(ctx, "Pause took %s", time.Since(start))
 	}()
-	if err := c.SaveSnapshot(ctx); err != nil {
-		log.CtxErrorf(ctx, "Error saving snapshot: %s", err)
+
+	snapshotType, memSnapshotName, baseSnapshotUnpackDir, err := c.snapshotDetails(ctx)
+	if err != nil {
 		return err
 	}
 
-	if err := c.Remove(ctx); err != nil {
-		log.CtxErrorf(ctx, "Error cleaning up after pause: %s", err)
+	if err = c.pauseVM(ctx); err != nil {
 		return err
 	}
+
+	// If an older snapshot is present -- nuke it since we're writing a new one.
+	if err = c.cleanupOldSnapshots(memSnapshotName, vmStateSnapshotName); err != nil {
+		return err
+	}
+
+	if err = c.createSnapshot(ctx, snapshotType, memSnapshotName, vmStateSnapshotName); err != nil {
+		return err
+	}
+
+	// Stop the VM and UFFD page fault handler to ensure nothing is modifying the snapshot files as we save them
+	if c.machine != nil {
+		// Note: we don't attempt any kind of clean shutdown here because we have already taken
+		// a snapshot
+		if err := c.machine.StopVMM(); err != nil {
+			log.CtxErrorf(ctx, "Error stopping VMM: %s", err)
+			return err
+		}
+		c.machine = nil
+	}
+	if c.uffdHandler != nil {
+		if err := c.uffdHandler.Stop(); err != nil {
+			log.CtxErrorf(ctx, "Error stopping uffd handler: %s", err)
+			return err
+		}
+		c.uffdHandler = nil
+	}
+
+	if err = c.saveSnapshot(ctx, snapshotType, baseSnapshotUnpackDir, memSnapshotName); err != nil {
+		return err
+	}
+
+	// Finish cleaning up VM resources
+	if err = c.Remove(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *FirecrackerContainer) snapshotDetails(ctx context.Context) (snapshotType string, memSnapshotName string, baseSnapshotUnpackDir string, err error) {
```

Comment:
uber nit: it might be cleaner to pack up the 3 return values into a struct, so it could just return `(*snapshotDetails, error)`

```suggestion
func (c *FirecrackerContainer) snapshotDetails(ctx context.Context) (snapshotType string, memSnapshotName string, baseSnapshotUnpackDir string, err error) {
```